### PR TITLE
Chore: update go version to 1.20.1

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### What this PR does
This PR is to update mimir go version to 1.20.1 and golangci-lint to version 1.51.2
#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -213,6 +213,7 @@ Querying with using `{__mimir_storage__="ephemeral"}` selector no longer works.
 * [CHANGE] Reduce period of health checks in connection pools for querier->store-gateway, ruler->ruler, and alertmanager->alertmanager clients to 10s. This reduces the time to fail a gRPC call when the remote stops responding. #3168
 * [CHANGE] Hide TSDB block ranges period config from doc and mark it experimental. #3518
 * [FEATURE] Alertmanager: added Discord support. #3309
+* [ENHANCEMENT] Go: update go go 1.20.1.
```
Comment (aknuds1) [inline CHANGELOG.md:216]: This is in the wrong place, it needs to be under "main / unreleased".

**Review by aknuds1 — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (pracucci) [reply CHANGELOG.md:216]: [nit] Also there's a little typo. Should be "update go to...".

File: mimir-build-image/Dockerfile [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 FROM k8s.gcr.io/kustomize/kustomize:v4.5.5 as kustomize
 FROM alpine/helm:3.8.2 as helm
-FROM golang:1.19.3-bullseye
+FROM golang:1.20.1
```
Comment (pracucci) [inline mimir-build-image/Dockerfile:8]: Let's be specific about which linux distribution to use. You will need to rebuild the docker image after this change.

```suggestion
FROM golang:1.20.1-bullseye
```

Comment (aknuds1) [reply mimir-build-image/Dockerfile:8]: Yep agree with @pracucci, there's no need to drop the distro specifier.

Comment (ying-jeanne) [reply mimir-build-image/Dockerfile:8]: my appologies to `go go` :)

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,7 +1,6 @@
 # Changelog
 
 ## main / unreleased
-
```
Comment (aknuds1) [inline CHANGELOG.md:4]: Can you remove this change please?

**Review by aknuds1 — COMMENTED** [review]
Looks pretty good to me, but have to look more closely. Please drop the change on line 4 of CHANGELOG.md though.

File: .github/workflows/helm-ci.yml [hunk 4]
```diff
@@ -18,10 +18,12 @@ jobs:
   conftest:
     runs-on: ubuntu-latest
     container:
-      image: grafana/mimir-build-image:pr3976-e7cae18e3
+      image: grafana/mimir-build-image:goupdate-72d66708c
     steps:
       - name: Checkout repository
         uses: actions/checkout@v2
+        run: git config --global --add safe.directory /__w/mimir/mimir
```
Comment (pstibrany) [inline .github/workflows/helm-ci.yml:25]: Why are we doing these changes? Seems unrelated to Go update.

File: .github/workflows/test-build-deploy.yml [hunk 5]
```diff
@@ -16,18 +16,19 @@ concurrency:
...
+job:
   lint:
     runs-on: ubuntu-latest
     container:
-      image: grafana/mimir-build-image:pr3976-e7cae18e3
+      image: grafana/mimir-build-image:goupdate-72d66708c
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
```
Comment (pstibrany) [inline .github/workflows/test-build-deploy.yml:26]: This seems unrelated to Go update, and should be done in a separate PR.

**Review by pstibrany — COMMENTED** [review]
I am concerned about changes unrelated to update of Go version. Why do we need them in this PR?

File: .github/workflows/helm-ci.yml (hunk 4 above)
Comment (ying-jeanne) [reply .github/workflows/helm-ci.yml:25]: The pipeline failed with dubious ownership in repository, i have seen this [Issue](https://github.com/actions/runner-images/issues/6775), that says the latest ubuntu image activate the security patch which cause that issue. Most of people solved issue by adding the git config command. But I wonder why other PR are not impacted

Comment (pstibrany) [reply .github/workflows/helm-ci.yml:25]: Is that related to using `actions/checkout@v3`? Can you please update this PR description why the additional changes are needed?

Comment (aknuds1) [reply .github/workflows/helm-ci.yml:25]: I think the error is triggered by the Go compiler `-buildvcs` flag, but I'm not sure if it changed in 1.20 or what.

Comment (aknuds1) [reply .github/workflows/helm-ci.yml:25]: I believe [this](https://github.com/actions/checkout/issues/1169) is also the root issue.

File: .github/workflows/test-build-deploy.yml [hunk 6]
```diff
@@ -16,14 +16,16 @@ concurrency:
   group: "${{ github.ref_name }}-${{ github.head_ref }}"
   cancel-in-progress: true
 
-jobs:
+job:
```
Comment (aknuds1) [inline .github/workflows/test-build-deploy.yml:19]: Can you undo this change? Maybe it's what causes the builds to not run:

File: .github/workflows/helm-ci.yml (hunk 4 above)
Comment (ying-jeanne) [reply .github/workflows/helm-ci.yml]: @pstibrany Sorry I move the PR to draft, the pipeline is still not passing, I am trying different things. Once it is passed I would summarise the findings and separate what is needed to change in github action and the go version update into 2 PRs with a better description.

File: .github/workflows/test-build-deploy.yml [hunk 7]
```diff
@@ -184,9 +196,9 @@ jobs:
       - name: Upgrade golang
         uses: actions/setup-go@v2
         with:
-          go-version: 1.19.3
+          go-version: 1.20.1
       - name: Checkout repository
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
```
Comment (pstibrany) [inline .github/workflows/test-build-deploy.yml:201]: Do we need to do `run: git config --global --add safe.directory /__w/mimir/mimir` magic here as well?

File: .github/workflows/test-build-deploy.yml [hunk 8]
```diff
@@ -228,10 +240,10 @@ jobs:
     if: (startsWith(github.ref, 'refs/tags/') || startsWith(github.ref, 'refs/heads/r') ) && github.event_name == 'push' && github.repository == 'grafana/mimir'
     runs-on: ubuntu-latest
     container:
-      image: grafana/mimir-build-image:pr3976-e7cae18e3
+      image: grafana/mimir-build-image:goupdate-751733fe1
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
```
Comment (pstibrany) [inline .github/workflows/test-build-deploy.yml:246]: Do we need to run `run: git config --global --add safe.directory /__w/mimir/mimir` here too?

File: mimir-build-image/Dockerfile [hunk 9]
```diff
@@ -4,8 +4,8 @@
 # Provenance-includes-copyright: The Cortex Authors.
 
 FROM k8s.gcr.io/kustomize/kustomize:v4.5.5 as kustomize
-FROM alpine/helm:3.8.2 as helm
-FROM golang:1.19.3-bullseye
+FROM alpine/helm:3.11.1 as helm
```
Comment (pstibrany) [inline mimir-build-image/Dockerfile:7]: We only copy Helm from this image (`COPY --from=helm /usr/bin/helm /usr/bin/helm` instruction). Do we need to update this image in this PR?

**Review by pstibrany — COMMENTED** [review]
Thank you.

Would you please update PR description with changes from this PR?

Please resolve conflicts with `main` branch, so that we can merge this.

File: mimir-build-image/Dockerfile (hunk 9 above)
Comment (aknuds1) [reply mimir-build-image/Dockerfile:7]: We don't need to upgrade Helm in this PR. I just suggested trying to upgrade it, after Krajo said an upgrade should be safe. At the beginning it wasn't clear if the build was failing due to Helm (the error was difficult to understand), so it was worth trying.

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (aknuds1) [reply .github/workflows/test-build-deploy.yml:201]: @pstibrany though evidently the build works, so I'm unsure whether or why we need `git config --global --add safe.directory` at all.

Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:201]: When I look at diff in this PR, I see that we run `git config --global --add safe.directory /__w/mimir/mimir` after some checkouts, but not everywhere. When do we need to do that and when can we skip that step?

Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:201]: > @pstibrany though evidently the build works, so I'm unsure whether or why we need `git config --global --add safe.directory` at all.

If it's not required, let's remove it :)

File: mimir-build-image/Dockerfile (hunk 9 above)
Comment (aknuds1) [reply mimir-build-image/Dockerfile:7]: @pstibrany my reply was basically to confirm that it's not necessary in this PR :)

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (aknuds1) [reply .github/workflows/test-build-deploy.yml:201]: @pstibrany I can try to experiment in a draft PR of mine, to see which changes are necessary.

Comment (aknuds1) [reply .github/workflows/test-build-deploy.yml:201]: From experimenting, I find that the checkout@v3 [`set-safe-directory` option](https://github.com/actions/checkout#usage) doesn't work. Maybe it configures the Git config of a different user than the build runs as? That's why Ying's workaround is necessary I would say.

File: .github/workflows/test-build-deploy.yml (hunk 8 above)
Comment (aknuds1) [reply .github/workflows/test-build-deploy.yml:246]: Since it doesn't fail for some reason, I would say it's not necessary?

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (aknuds1) [reply .github/workflows/test-build-deploy.yml:201]: > Do we need to do run: git config --global --add safe.directory /__w/mimir/mimir magic here as well?

Since it doesn't fail (for some reason), I would say it's not necessary?

**Comment by pstibrany** [comment]
I've spent some time investigating `detected dubious ownership in repository` error.

This error is reported by Git when checked-out Git repository has different owner than current user.

When Github Action runs, it uses user id 1001. This user is also owner of `$RUNNER_WORKSPACE` and `$GITHUB_WORKSPACE` directories.

When action uses a container, all steps for given action are executed from inside the container. This includes "checkout" action or any command exection. Files created by container has `root` owner.

`checkout` actions checks out Mimir repositor directly into `$GITHUB_WORKSPACE`, with no extra directory in between.

This creates a problem -- `$GITHUB_WORKSPACE` has owner 1001, but files in it have `root` owner. Git from version [v2.30.3](https://github.com/git/git/commit/8959555cee7ec045958f9b6dd62e541affb7e7d9) doesn't like that.

While update of the build image in this PR apparently didn't change Git version (`git --version` reports `v2.30.2`), this is deceiving, because Debian maintainers [backported](https://release.debian.org/proposed-updates/bullseye_diffs/git_2.30.2-1+deb11u1.debdiff) ownership checks into this older version. That's why we started to see this problem in this PR.

(Changes in Debian git packages: https://tracker.debian.org/pkg/git)

File: .github/workflows/helm-ci.yml [hunk 10]
```diff
@@ -18,11 +18,14 @@ jobs:
...
     runs-on: ubuntu-latest
     container:
-      image: grafana/mimir-build-image:pr3976-e7cae18e3
+      image: grafana/mimir-build-image:goupdate-751733fe1
     steps:
       - name: Checkout repository
         uses: actions/checkout@v2
 
+      - name: Run Git Config
+        run: git config --global --add safe.directory /__w/mimir/mimir
```
Comment (pstibrany) [inline .github/workflows/helm-ci.yml:27]:
```suggestion
        run: git config --global --add safe.directory '*'
```

We can use this version and not rely on specific placement of Mimir repository. (We could use `$GITHUB_WORKSPACE` too)

Applies to all these lines.

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:201]: > Do we need to do `run: git config --global --add safe.directory /__w/mimir/mimir` magic here as well?

To answer my question, we only need it when we want to run `git` command from our steps, and action runs inside the container. So not needed here.

**Comment by pstibrany** [comment]
I've pushed small changes to the PR to minimize diff, and suggest to merge as-is, if checks pass. Marked as ready for review.

**Review by pstibrany — APPROVED** [review]
